torch/csrc/utils/python_arg_parser.cpp

```diff
@@ -320,53 +320,47 @@ bool FunctionSignature::parse(PyObject* args, PyObject* kwargs, PyObject* dst[],
   int i = 0;
   for (auto& param : params) {
     PyObject* obj = nullptr;
+    bool is_kwd = false;
     if (arg_pos < nargs) {
       obj = PyTuple_GET_ITEM(args, arg_pos);
-      if (obj == Py_None && param.allow_none) {
-        dst[i++] = nullptr;
-        arg_pos++;
-        continue;
-      }
-      if (param.check(obj)) {
-        dst[i++] = obj;
-        arg_pos++;
-        continue;
-      }
-      if (allow_varargs_intlist && arg_pos == 0) {
-        dst[i++] = args;
-        arg_pos += nargs;
-        continue;
-      }
+    } else if (kwargs) {
+      obj = PyDict_GetItem(kwargs, param.python_name);
+      is_kwd = true;
+    }
+
+    if ((!obj && param.optional) || (obj == Py_None && param.allow_none)) {
+      dst[i++] = nullptr;
+    } else if (!obj) {
       if (raise_exception) {
+        // foo() missing 1 required positional argument: "b"
+        missing_args(*this, i);
+      }
+      return false;
+    } else if (param.check(obj)) {
+      dst[i++] = obj;
+    } else if (allow_varargs_intlist && arg_pos == 0 && !is_kwd) {
+      dst[i++] = args;
+      arg_pos = nargs;
```

Comment:
nit: this will end up being 1 past the length because it will be incremented again.  Probably not worth complicating the code, though.